src/coreclr/src/utilcode/pedecoder.cpp

```diff
@@ -2319,22 +2319,70 @@ READYTORUN_HEADER * PEDecoder::FindReadyToRunHeader() const
     }
     CONTRACTL_END;
 
-    IMAGE_DATA_DIRECTORY *pDir = &GetCorHeader()->ManagedNativeHeader;
-
-    if (VAL32(pDir->Size) >= sizeof(READYTORUN_HEADER) && CheckDirectory(pDir))
+    if (HasCorHeader())
     {
-        PTR_READYTORUN_HEADER pHeader = PTR_READYTORUN_HEADER((TADDR)GetDirectoryData(pDir));
-        if (pHeader->Signature == READYTORUN_SIGNATURE)
+        IMAGE_DATA_DIRECTORY* pDir = &GetCorHeader()->ManagedNativeHeader;
+
+        if (VAL32(pDir->Size) >= sizeof(READYTORUN_HEADER) && CheckDirectory(pDir))
         {
-            const_cast<PEDecoder *>(this)->m_pReadyToRunHeader = pHeader;
-            return pHeader;
+            PTR_READYTORUN_HEADER pHeader = PTR_READYTORUN_HEADER((TADDR)GetDirectoryData(pDir));
+            if (pHeader->Signature == READYTORUN_SIGNATURE)
+            {
+                const_cast<PEDecoder*>(this)->m_pReadyToRunHeader = pHeader;
+                return pHeader;
+            }
         }
     }
+    
+    READYTORUN_HEADER *nativeReadyToRunHeader = FindNativeReadyToRunHeader();
+    if (nativeReadyToRunHeader != NULL)
+    {
+        return nativeReadyToRunHeader;
+    }
 
     const_cast<PEDecoder *>(this)->m_flags |= FLAG_HAS_NO_READYTORUN_HEADER;
     return NULL;
 }
 
+READYTORUN_HEADER *PEDecoder::FindNativeReadyToRunHeader() const
+{
+#ifndef DACCESS_COMPILE
+    // Get the export directory entry
+    PIMAGE_DATA_DIRECTORY exportDirectoryEntry = GetDirectoryEntry(IMAGE_DIRECTORY_ENTRY_EXPORT);
+    if (exportDirectoryEntry->VirtualAddress == 0 || exportDirectoryEntry->Size == 0)
+    {
+        return NULL;
+    }
+    
+    const uint8_t *imageBase = (const uint8_t *)GetBase();
+    const uint8_t *exportDirectory = &imageBase[exportDirectoryEntry->VirtualAddress];
+    
+    // TODO: endianness?
+    // +0x18: number of name pointers
+    int namePointerCount = *(const int32_t *)&exportDirectory[0x18];
```

Comment:
Are these bounds checked somewhere else?


Is `imageBase` supposed to be aligned on a int32_t?  Because if not, you might want to define a helper like this:

```C
static inline int32_t ReadLEInt32(const uint8_t *base, ptrdiff_t offset)
{
    int32_t value;

    memcpy(&value, base + offset, sizeof(value));
    return le32toh(value);
}
```

Which takes care of this not causing an unaligned load, and converts the little endian encoding to whatever the current CPU uses.

Reply:
I have added a simple helper ReadUInt32 to pedecoder.cpp to combine the range check, unaligned read and little endian decoding. Thanks for keeping an eye on robustness of our runtime!